SSWC/doc/man/man3/App.3
.TH "App" 3 "Mon May 9 2016" "Version 0.1" "MissionsVisualizer" \" -*- nroff -*-
.ad l
.nh
.SH NAME
App \- 
.SH SYNOPSIS
.br
.PP
.PP
\fC#include <App\&.h>\fP
.SS "Classes"

.in +1c
.ti -1c
.RI "struct \fBDefaultUnits\fP"
.br
.in -1c
.SS "Public Types"

.in +1c
.ti -1c
.RI "enum \fBUnitsType\fP { \fBUT_DEFAULT\fP, \fBUT_METRIC\fP, \fBUT_IMPERIAL\fP }"
.br
.in -1c
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBApp\fP ()"
.br
.ti -1c
.RI "\fB~App\fP ()"
.br
.ti -1c
.RI "\fBvoid\fP \fBInit\fP ()"
.br
.ti -1c
.RI "\fBvoid\fP \fBLoadKernel\fP (const \fBstd::string\fP &file) const "
.br
.ti -1c
.RI "\fBvoid\fP \fBSetLoggingFile\fP (const \fBstd::string\fP &file) const "
.br
.ti -1c
.RI "\fBvoid\fP \fBSetReferenceFrame\fP (const \fBFrame\fP &\fBref\fP)"
.br
.ti -1c
.RI "const \fBFrame\fP & \fBGetReferenceFrame\fP () const "
.br
.ti -1c
.RI "\fBvoid\fP \fBSetDefaultUnits\fP (\fBUnitsType\fP \fBunits\fP)"
.br
.ti -1c
.RI "\fBvoid\fP \fBSetIndividualUnits\fP (const \fBDefaultUnits\fP &defUnits)"
.br
.ti -1c
.RI "const \fBLengthUnit\fP & \fBLengthUnit\fP () const "
.br
.ti -1c
.RI "const \fBVelocityUnit\fP & \fBVelocityUnit\fP () const "
.br
.ti -1c
.RI "const \fBMassUnit\fP & \fBMassUnit\fP () const "
.br
.ti -1c
.RI "const \fBGravitationalParameterUnit\fP & \fBGMUnit\fP () const "
.br
.ti -1c
.RI "const \fBAccelerationUnit\fP & \fBAccelerationUnit\fP () const "
.br
.ti -1c
.RI "\fBvoid\fP \fBLoadChildren\fP (const \fBSpaceObject\fP &parent, bool includeSelf=false, bool recursive=false)"
.br
.ti -1c
.RI "\fBvoid\fP \fBLoadParent\fP (const \fBSpaceObject\fP &child, bool includeMassCenter, bool includeSelf=false, bool recursive=false)"
.br
.ti -1c
.RI "\fBvoid\fP \fBLoadMoons\fP (const \fBSpaceObject\fP &parent)"
.br
.ti -1c
.RI "\fBvoid\fP \fBLoadMoons\fP (const \fBstd::string\fP &parentName)"
.br
.ti -1c
.RI "\fBvoid\fP \fBLoadSolarSystem\fP (bool entire=false)"
.br
.ti -1c
.RI "\fBvoid\fP \fBLoadAllAvailableObjects\fP ()"
.br
.ti -1c
.RI "\fBvoid\fP \fBAddObject\fP (const \fBSpaceObject\fP &\fBobj\fP)"
.br
.ti -1c
.RI "size_t \fBGetObjectsLength\fP () const "
.br
.ti -1c
.RI "\fBSpaceObject\fP & \fBGetObjectByIndex\fP (size_t idx)"
.br
.ti -1c
.RI "bool \fBCheckObjectExists\fP (long \fBid\fP)"
.br
.ti -1c
.RI "bool \fBCheckObjectExists\fP (const \fBstd::string\fP &\fBname\fP)"
.br
.ti -1c
.RI "bool \fBCheckObjectExists\fP (const \fBSpaceObject\fP &sample)"
.br
.ti -1c
.RI "\fBSpaceObject\fP & \fBRetrieveObject\fP (long \fBid\fP)"
.br
.ti -1c
.RI "\fBSpaceObject\fP & \fBRetrieveObject\fP (const \fBstd::string\fP &\fBname\fP)"
.br
.ti -1c
.RI "\fBSpaceObject\fP & \fBRetrieveObject\fP (const \fBSpaceObject\fP &sample)"
.br
.ti -1c
.RI "std::vector< \fBSpaceObject\fP * > \fBGetLoadedPlanets\fP ()"
.br
.ti -1c
.RI "std::vector< \fBSpaceObject\fP * > \fBGetLoadedMoonsOf\fP (const \fBSpaceObject\fP &planet)"
.br
.ti -1c
.RI "std::vector< \fBSpaceObject\fP * > \fBGetLoadedMoonsOf\fP (const \fBstd::string\fP &\fBplanetName\fP)"
.br
.ti -1c
.RI "std::vector< \fBSpaceObject\fP * > \fBGetLoadedMoons\fP ()"
.br
.ti -1c
.RI "std::vector< \fBSpaceObject\fP * > \fBGetLoadedBarycenters\fP ()"
.br
.in -1c
.SH "Detailed Description"
.PP 
Definition at line 13 of file App\&.h\&.
.SH "Member Enumeration Documentation"
.PP 
.SS "enum \fBApp::UnitsType\fP"

.PP
\fBEnumerator\fP
.in +1c
.TP
\fB\fIUT_DEFAULT \fP\fP
.TP
\fB\fIUT_METRIC \fP\fP
.TP
\fB\fIUT_IMPERIAL \fP\fP
.PP
Definition at line 16 of file App\&.h\&.
.SH "Constructor & Destructor Documentation"
.PP 
.SS "App::App ()"

.PP
Definition at line 3 of file App\&.cpp\&.
.SS "App::~App ()"

.PP
Definition at line 8 of file App\&.cpp\&.
.SH "Member Function Documentation"
.PP 
.SS "const \fBAccelerationUnit\fP & App::AccelerationUnit () const"

.PP
Definition at line 90 of file App\&.cpp\&.
.SS "\fBvoid\fP App::AddObject (const \fBSpaceObject\fP &obj)"

.PP
Definition at line 199 of file App\&.cpp\&.
.SS "bool App::CheckObjectExists (longid)"

.PP
Definition at line 218 of file App\&.cpp\&.
.SS "bool App::CheckObjectExists (const \fBstd::string\fP &name)"

.PP
Definition at line 229 of file App\&.cpp\&.
.SS "bool App::CheckObjectExists (const \fBSpaceObject\fP &sample)"

.PP
Definition at line 240 of file App\&.cpp\&.
.SS "std::vector< \fBSpaceObject\fP * > App::GetLoadedBarycenters ()"

.PP
Definition at line 318 of file App\&.cpp\&.
.SS "std::vector< \fBSpaceObject\fP * > App::GetLoadedMoons ()"

.PP
Definition at line 306 of file App\&.cpp\&.
.SS "std::vector< \fBSpaceObject\fP * > App::GetLoadedMoonsOf (const \fBSpaceObject\fP &planet)"

.PP
Definition at line 286 of file App\&.cpp\&.
.SS "std::vector< \fBSpaceObject\fP * > App::GetLoadedMoonsOf (const \fBstd::string\fP &planetName)"

.PP
Definition at line 301 of file App\&.cpp\&.
.SS "std::vector< \fBSpaceObject\fP * > App::GetLoadedPlanets ()"

.PP
Definition at line 274 of file App\&.cpp\&.
.SS "\fBSpaceObject\fP & App::GetObjectByIndex (size_tidx)"

.PP
Definition at line 213 of file App\&.cpp\&.
.SS "size_t App::GetObjectsLength () const"

.PP
Definition at line 208 of file App\&.cpp\&.
.SS "const \fBFrame\fP & App::GetReferenceFrame () const"

.PP
Definition at line 40 of file App\&.cpp\&.
.SS "const \fBGravitationalParameterUnit\fP & App::GMUnit () const"

.PP
Definition at line 85 of file App\&.cpp\&.
.SS "\fBvoid\fP App::Init ()"

.PP
Definition at line 19 of file App\&.cpp\&.
.SS "const \fBLengthUnit\fP & App::LengthUnit () const"

.PP
Definition at line 70 of file App\&.cpp\&.
.SS "\fBvoid\fP App::LoadAllAvailableObjects ()"

.PP
Definition at line 185 of file App\&.cpp\&.
.SS "\fBvoid\fP App::LoadChildren (const \fBSpaceObject\fP &parent, boolincludeSelf = \fCfalse\fP, boolrecursive = \fCfalse\fP)"

.PP
Definition at line 100 of file App\&.cpp\&.
.SS "\fBvoid\fP App::LoadKernel (const \fBstd::string\fP &file) const"

.PP
Definition at line 25 of file App\&.cpp\&.
.SS "\fBvoid\fP App::LoadMoons (const \fBSpaceObject\fP &parent)"

.PP
Definition at line 150 of file App\&.cpp\&.
.SS "\fBvoid\fP App::LoadMoons (const \fBstd::string\fP &parentName)"

.PP
Definition at line 160 of file App\&.cpp\&.
.SS "\fBvoid\fP App::LoadParent (const \fBSpaceObject\fP &child, boolincludeMassCenter, boolincludeSelf = \fCfalse\fP, boolrecursive = \fCfalse\fP)"

.PP
Definition at line 123 of file App\&.cpp\&.
.SS "\fBvoid\fP App::LoadSolarSystem (boolentire = \fCfalse\fP)"

.PP
Definition at line 165 of file App\&.cpp\&.
.SS "const \fBMassUnit\fP & App::MassUnit () const"

.PP
Definition at line 80 of file App\&.cpp\&.
.SS "\fBSpaceObject\fP & App::RetrieveObject (longid)"

.PP
Definition at line 245 of file App\&.cpp\&.
.SS "\fBSpaceObject\fP & App::RetrieveObject (const \fBstd::string\fP &name)"

.PP
Definition at line 258 of file App\&.cpp\&.
.SS "\fBSpaceObject\fP & App::RetrieveObject (const \fBSpaceObject\fP &sample)"

.PP
Definition at line 269 of file App\&.cpp\&.
.SS "\fBvoid\fP App::SetDefaultUnits (\fBUnitsType\fPunits)"

.PP
Definition at line 45 of file App\&.cpp\&.
.SS "\fBvoid\fP App::SetIndividualUnits (const \fBDefaultUnits\fP &defUnits)"

.PP
Definition at line 95 of file App\&.cpp\&.
.SS "\fBvoid\fP App::SetLoggingFile (const \fBstd::string\fP &file) const"

.PP
Definition at line 30 of file App\&.cpp\&.
.SS "\fBvoid\fP App::SetReferenceFrame (const \fBFrame\fP &ref)"

.PP
Definition at line 35 of file App\&.cpp\&.
.SS "const \fBVelocityUnit\fP & App::VelocityUnit () const"

.PP
Definition at line 75 of file App\&.cpp\&.

.SH "Author"
.PP 
Generated automatically by Doxygen for MissionsVisualizer from the source code\&.


SSWC/src/App.cpp
#include "App.h"

App::App() : refFrame(REF_FRAME_DEFAULT)
{
	Init();
}

App::~App()
{
	for(size_t i = 0; i < objects.size(); i++)
	{
		delete objects[i];
		objects[i] = nullptr;
	}

	objects.clear();
}

void App::Init()
{
	CSpiceUtil::SetErrorHandlingParams("return", "null");
	SetDefaultUnits(UT_DEFAULT);
}

void App::LoadKernel(const std::string& file) const
{
	CSpiceUtil::LoadKernel(file);
}

void App::SetLoggingFile(const std::string& file) const
{
	CSpiceUtil::SetLoggingFile(file);
}

void App::SetReferenceFrame(const Frame& ref)
{
	refFrame = ref;
}

const Frame& App::GetReferenceFrame() const
{
	return refFrame;
}

void App::SetDefaultUnits(UnitsType units)
{
	this->units.accUnit = Units::Metric::ms2;
	this->units.gmUnit = Units::Metric::gm;

	switch(units)
	{
	case UT_DEFAULT:
		this->units.lengthUnit = Units::Metric::kilometers;
		this->units.velocityUnit = Units::Metric::kmps;
		this->units.massUnit = Units::Metric::kilograms;
		break;
	case UT_IMPERIAL:
		this->units.lengthUnit = Units::Imperial::miles;
		this->units.velocityUnit = Units::Imperial::mph;
		this->units.massUnit = Units::Imperial::pounds;
		break;
	case UT_METRIC:
		this->units.lengthUnit = Units::Metric::meters;
		this->units.velocityUnit = Units::Metric::mps;
		this->units.massUnit = Units::Metric::kilograms;
		break;
	}
}

const LengthUnit& App::LengthUnit() const
{
	return units.lengthUnit;
}

const VelocityUnit& App::VelocityUnit() const
{
	return units.velocityUnit;
}

const MassUnit& App::MassUnit() const
{
	return units.massUnit;
}

const GravitationalParameterUnit& App::GMUnit() const
{
	return units.gmUnit;
}

const AccelerationUnit& App::AccelerationUnit() const
{
	return units.accUnit;
}

void App::SetIndividualUnits(const DefaultUnits& defUnits)
{
	std::memcpy(&this->units, &defUnits, sizeof(DefaultUnits));
}

void App::LoadChildren(const SpaceObject& parent, bool includeSelf, bool recursive)
{
	long parentId = parent.GetSpiceId();

	std::vector<long> childrenIds = SpaceObject::FindChildObjectIds(parentId);

	if(includeSelf)
		AddObject(parent);

	for(size_t i = 0; i < childrenIds.size(); i++)
	{
		long childId = childrenIds[i];

		if(SpaceObject::IsBody(childId))
			AddObject(SpaceBody(childId));
		else
			AddObject(SpaceObject(childId));

		if(recursive)
			LoadChildren(SpaceObject(childId), false, true);
	}
}

void App::LoadParent(const SpaceObject& child, bool includeMassCenter, bool includeSelf, bool recursive)
{
	if(includeSelf)
		AddObject(child);

	long childId = child.GetSpiceId();
	long parentId = SpaceObject::FindParentObjectId(childId);

	if(childId == parentId)
		return;

	if(SpaceObject::IsBody(parentId))
		AddObject(SpaceBody(parentId));
	else
		AddObject(SpaceObject(parentId));

	if(includeMassCenter)
	{
		long massCenterId = SpaceObject::FindChildMassCenterId(parentId);
		if(massCenterId != parentId)
			AddObject(SpaceBody(massCenterId));
	}

	if(recursive)
		LoadParent(SpaceObject(parentId), includeMassCenter, false, true);
}

void App::LoadMoons(const SpaceObject& planet)
{
	long barycenterId = SpaceObject::FindParentObjectId(planet.GetSpiceId());

	if(!SpaceObject::IsPlanetaryBarycenter(barycenterId))
		CSpiceUtil::SignalError("LoadMoons failed: 'planet' argument is not a valid planet object");

	return LoadChildren(SpaceObject(barycenterId), false, false);
}

void App::LoadMoons(const std::string& parentName)
{
	return LoadMoons(SpaceObject(parentName));
}

void App::LoadSolarSystem(bool entire)
{
	if(!entire)
	{
		std::vector<long> barycenterIds = SpaceObject::FindChildObjectIds(SpaceObject::SSB.GetSpiceId());

		for(size_t i = 0; i < barycenterIds.size(); i++)
		{
			long planetId = barycenterIds[i] * 100 + 99;

			if(SpaceObject::IsPlanet(planetId) || SpaceObject::IsSun(planetId))
				AddObject(SpaceBody(planetId));
		}
	}
	else
	{
		LoadChildren(SpaceObject::SSB, true, true);
	}
}

void App::LoadAllAvailableObjects()
{
	std::vector<long> loadedIds = SpaceObject::GetLoadedSpkIds();
	for(size_t i = 0; i < loadedIds.size(); i++)
	{
		if(SpaceObject::IsBody(loadedIds[i]))
			AddObject(SpaceBody(loadedIds[i]));
		else if(SpaceObject::ValidateId(loadedIds[i]))
			AddObject(SpaceObject(loadedIds[i]));
	}

	AddObject(SpaceObject::SSB);
}

void App::AddObject(const SpaceObject& obj)
{
	if(CheckObjectExists(obj) == false)
	{
		objects.push_back(obj.Clone());
		LoadParent(obj, true, false, true);
	}
}

size_t App::GetObjectsLength() const
{
	return objects.size();
}

SpaceObject& App::GetObjectByIndex(size_t idx)
{
	return *objects[idx];
}

bool App::CheckObjectExists(long id)
{
	for(size_t i = 0; i < objects.size(); i++)
	{
		if(objects[i]->GetSpiceId() == id)
			return true;
	}

	return false;
}

bool App::CheckObjectExists(const std::string& name)
{
	for(size_t i = 0; i < objects.size(); i++)
	{
		if(objects[i]->GetName() == name)
			return true;
	}

	return CheckObjectExists(SpaceObject(name));
}

bool App::CheckObjectExists(const SpaceObject& sample)
{
	return CheckObjectExists(sample.GetSpiceId());
}

SpaceObject& App::RetrieveObject(long id)
{
	for(size_t i = 0; i < objects.size(); i++)
	{
		if(objects[i]->GetSpiceId() == id)
			return GetObjectByIndex(i);
	}

	CSpiceUtil::SignalError("RetrieveObject haven't found requested object");

	throw;
}

SpaceObject& App::RetrieveObject(const std::string& name)
{
	for(size_t i = 0; i < objects.size(); i++)
	{
		if(objects[i]->GetName() == name)
			return GetObjectByIndex(i);
	}

	return RetrieveObject(SpaceObject(name));
}

SpaceObject& App::RetrieveObject(const SpaceObject& sample)
{
	return RetrieveObject(sample.GetSpiceId());
}

std::vector<SpaceObject*> App::GetLoadedPlanets()
{
	std::vector<SpaceObject*> planets;
	for(size_t i = 0; i < objects.size(); i++)
	{
		if(objects[i]->IsPlanet())
			planets.push_back(objects[i]);
	}

	return planets;
}

std::vector<SpaceObject*> App::GetLoadedMoonsOf(const SpaceObject& planet)
{
	std::vector<SpaceObject*> moons;

	std::vector<long> allMoonIds = SpaceObject::FindMoonIds(planet.GetSpiceId());

	for(size_t i = 0; i < allMoonIds.size(); i++)
	{
		if(CheckObjectExists(allMoonIds[i]))
			moons.push_back(&RetrieveObject(allMoonIds[i]));
	}

	return moons;
}

std::vector<SpaceObject*> App::GetLoadedMoonsOf(const std::string& planetName)
{
	return GetLoadedMoonsOf(SpaceObject(planetName));
}

std::vector<SpaceObject*> App::GetLoadedMoons()
{
	std::vector<SpaceObject*> moons;
	for(size_t i = 0; i < objects.size(); i++)
	{
		if(objects[i]->IsMoon())
			moons.push_back(objects[i]);
	}

	return moons;
}

std::vector<SpaceObject*> App::GetLoadedBarycenters()
{
	std::vector<SpaceObject*> barycenters;
	for(size_t i = 0; i < objects.size(); i++)
	{
		if(objects[i]->IsBarycenter())
			barycenters.push_back(objects[i]);
	}

	return barycenters;
}
